modules/planning/tasks/deciders/path_assessment_decider.cc

```diff
@@ -76,30 +75,69 @@ Status PathAssessmentDecider::Process(
       continue;
     }
     ADEBUG << "Path length = " << curr_path_data.frenet_frame_path().size();
     SetPathInfo(*reference_line_info, &curr_path_data);
+    // Trim all the lane-borrowing paths so that it ends with an in-lane
+    // position.
+    TrimTailingOutLanePoints(&curr_path_data);
   }
 
-  // Pick the optimal path
+  // 3. Pick the optimal path.
   std::sort(valid_path_data.begin(), valid_path_data.end(),
             [](const PathData& lhs, const PathData& rhs) {
-              if (lhs.path_label() != rhs.path_label()) {
-                return lhs.path_label() == "regular";
+              // Regular path goes before fallback path.
+              if (lhs.path_label().find("regular") !=
+                  rhs.path_label().find("regular")) {
+                return lhs.path_label().find("regular") != std::string::npos;
+              }
+              // Select longer path.
+              constexpr double kPathLengthComparisonTolerance = 5.0;
+              if (std::fabs(lhs.discretized_path().back().s() -
+                            rhs.discretized_path().back().s()) >
+                  kPathLengthComparisonTolerance) {
+                return lhs.discretized_path().size() <
+                       rhs.discretized_path().size();
+              }
+              // If roughly same length,
+              // then select self-lane path over borrowed-lane paths.
+              if (lhs.path_label().find("self") != std::string::npos ||
+                  rhs.path_label().find("self") != std::string::npos) {
+                return lhs.path_label().find("self") != std::string::npos;
+              }
+              // If roughly same length, and must borrow neighbor lane,
+              // then prefer to borrow forward lane rather than reverse lane.
+              if (ContainsOutOnReverseLane(lhs.path_point_decision_guide()) !=
+                  ContainsOutOnReverseLane(rhs.path_point_decision_guide())) {
+                return !ContainsOutOnReverseLane(
+                           lhs.path_point_decision_guide());
               }
-              // TODO(jiacheng): update this to include more lane-info
-              // in sorting.
-              return lhs.path_label() == "regular";
+              // If same length, both neighbor lane are forward,
+              // then select the one that returns back to in-lane earlier.
+              if (GetBackToInLaneIndex(lhs.path_point_decision_guide()) !=
+                  GetBackToInLaneIndex(rhs.path_point_decision_guide())) {
+                return GetBackToInLaneIndex(lhs.path_point_decision_guide()) <
+                       GetBackToInLaneIndex(rhs.path_point_decision_guide());
+              }
+              // If same length, both forward, back to inlane at same time,
+              // select the left one to side-pass.
+              if (lhs.path_label().find("left") !=
+                  rhs.path_label().find("left")) {
```

Comment:
Fit in one line?

Reply:
Not able to fit in one line due to over-length but modified in another way.